### DISCUSSION TITLE
Tailwind CSS IntelliSense and Twin

### DISCUSSION THREAD

**Discussion by ben-rogerson** [ben-rogerson/twin.macro#227] [General]
<a href="https://marketplace.visualstudio.com/items?itemName=bradlc.vscode-tailwindcss"><img src="https://user-images.githubusercontent.com/21288568/100925143-87073c00-3531-11eb-9b74-50529acfcd57.png" alt="logo" width="360"></a>

Good news, Brad has just pushed a new release that includes an [experimental custom regex feature](https://github.com/tailwindlabs/tailwindcss-intellisense/issues/129#issuecomment-735915659).

## Installation

Install the extension in VSCode:

<img src="https://user-images.githubusercontent.com/21288568/100925086-6fc84e80-3531-11eb-8d13-8658c8c41b62.png" alt="logo" width="360">

From your command palette select: "Preferences: Open Settings (JSON)" and add this to your settings:

```js
"scss.validate": false,
"editor.quickSuggestions": {
    "strings": true
},
"editor.autoClosingQuotes": "always",
"tailwindCSS.experimental.classRegex": [
    "tw`([^`]*)", // tw`...`
    "tw=\"([^\"]*)", // <div tw="..." />
    "tw={\"([^\"}]*)", // <div tw={"..."} />
    "tw\\.\\w+`([^`]*)", // tw.xxx`...`
    "tw\\(.*?\\)`([^`]*)" // tw(Component)`...`
],
"tailwindCSS.includeLanguages": {
  "typescript": "javascript",
  "typescriptreact": "javascript"
}
```

Note: If you use single quotes in your project then replace them in the regex sets above.

In order for the extension to activate you must have tailwindcss installed and a Tailwind config file named `tailwind.config.js` or `tailwind.js` in your workspace.

You'll also need to install PostCSS in your project:

```shell
npm install -D postcss 
yarn add postcss -D
```

You'll then see class auto-completions from your merged tailwind.config

![image](https://user-images.githubusercontent.com/21288568/102535974-e00dcd00-40f8-11eb-80a9-c383de3339cd.png)

**Comment by lightyen** [comment]


**Reply by ben-rogerson** [reply]
I'll add this feature in the next version #253

**Reply by ben-rogerson** [reply]
Your own extension? sounds interesting!

**Reply by lightyen** [reply]
> Your own extension? sounds interesting!

https://github.com/lightyen/vscode-tailwindcss-twin

**Comment by melloc01** [comment]
@ben-rogerson, please add one more case: ` tw(Component)`` `

```
  "tailwindCSS.experimental.classRegex": [
    "tw`([^`]*)", // tw`...`
    "tw=\"([^\"]*)", // <div tw="..." />
    "tw={\"([^\"}]*)", // <div tw={"..."} />
    "tw\\.\\w+`([^`]*)", // tw.xxx`...`
    "tw\(.*?\)`([^`]*)" // tw(Component)`...`
  ]
```

**Reply by ben-rogerson** [reply]
Done, thank you

**Comment by matepaiva** [comment]
Is there anyway to show an error in vscode linter, as it already shows in the console and in the browser, when a wrong class is added to twin?

**Reply by ben-rogerson** [reply]
Unfortunately the vscode plugin has no way of knowing which suggestions twin is making.

**Comment by kelvindecosta** [comment]
You can also use the same rules with [Headwind](https://github.com/heybourn/headwind):

```json
{
"headwind.classRegex": {
    // ...
    "javascript": "tw`([^`]*)|tw=\"([^\"]*)|tw={\"([^\"}]*)|tw\\.\\w+`([^`]*)|tw\\(.*?\\)`([^`]*)|(?:\\bclassName\\s*=\\s*[\\\"\\']([_a-zA-Z0-9\\s\\-\\:\\/]+)[\\\"\\'])|(?:\\btw\\s*`([_a-zA-Z0-9\\s\\-\\:\\/]*)`)",
    "javascriptreact": "tw`([^`]*)|tw=\"([^\"]*)|tw={\"([^\"}]*)|tw\\.\\w+`([^`]*)|tw\\(.*?\\)`([^`]*)|(?:\\bclassName\\s*=\\s*[\\\"\\']([_a-zA-Z0-9\\s\\-\\:\\/]+)[\\\"\\'])|(?:\\btw\\s*`([_a-zA-Z0-9\\s\\-\\:\\/]*)`)",
  }
}
```

**Comment by pprg1996** [comment]
Hello, Ben.
There's a missing part of the instructions. There has to be a "tailwind.config.js" file in the root or else the extension won't work.

**Comment by JimmyLv** [comment]
How about WebStorm?

**Reply by ben-rogerson** [reply]
There is no plugin for WebStorm - sorry!

**Reply by dmwyatt** [reply]
As a stupid workaround, I usually use `className` so that I can get completion for tailwind class names (since WS will pull those out of the css file) and then switch those over to `tw`.  

I don't think it'd be difficult to write a plugin for Jetbrains IDEs, but I don't like writing Java or Kotlin and don't have the time right now anyway.

**Reply by SnowSuno** [reply]
It is now available by adding the config under `Settings > Languages & Frameworks > Style Sheets > Tailwind CSS` in versions after `2023.1`. [youtrack issue](https://youtrack.jetbrains.com/issue/WEB-48505)

**Comment by alanontheweb** [comment]
@ben-rogerson @lightyen  is there any way for the `vs-code-tailwind-twin` intellisense to work with the babel-plugin-twin (https://github.com/ben-rogerson/babel-plugin-twin)?

It only seems to autocomplete classes when I manually import the macro. 

Also for some reason, the classes don't have a green color when using `<div tw="border-2 flex">test</div>`

![image](https://user-images.githubusercontent.com/302498/111058550-3b273580-8444-11eb-9f3d-71230b33a028.png)

**Reply by ben-rogerson** [reply]
Lightyen has done some work here - [have a read of this issue](https://github.com/lightyen/vscode-tailwindcss-twin/issues/6)

**Comment by alanontheweb** [comment]
@ben-rogerson awesome, it works great!

**Comment by locona** [comment]
@ben-rogerson 
I'm currently developing using twin.macro with this configuration
In the following case, completion does not work.

The reason why I don't use extention is that every time I make a typo, completion runs and slows things down.


ref: https://github.com/ben-rogerson/twin.macro/discussions/227#discussion-38611
```
"scss.validate": false,
"editor.quickSuggestions": {
    "strings": true
},
"editor.autoClosingQuotes": "always",
"tailwindCSS.experimental.classRegex": [
    "tw`([^`]*)", // tw`...`
    "tw=\"([^\"]*)", // <div tw="..." />
    "tw={\"([^\"}]*)", // <div tw={"..."} />
    "tw\\.\\w+`([^`]*)", // tw.xxx`...`
    "tw\\(.*?\\)`([^`]*)" // tw(Component)`...`
],
"tailwindCSS.includeLanguages": {
  "typescript": "javascript",
  "typescriptreact": "javascript"
}
```

**Reply by ben-rogerson** [reply]
Have you tried the other extension - [vscode-tailwindcss-twin](https://github.com/lightyen/vscode-tailwindcss-twin)? Completion may be faster.

**Reply by locona** [reply]
@ben-rogerson 
yes !  The following refers to vscode-tailwindcss-twin.

> The reason why I don't use extention is that every time I make a typo, completion runs and slows things down.

As shown in the GIF below, the cursor movement  will be heavy.
![CleanShot 2021-12-12 at 13 40 49](https://user-images.githubusercontent.com/18649881/145700518-7253f694-a841-4475-844f-9af3f770943a.gif)

I've never done any extension development, so I don't know how to identify the cause.
I hope you can help me.

**Reply by ben-rogerson** [reply]
I've moved this to an issue in [lightyen/vscode-tailwindcss-twin](https://github.com/lightyen/vscode-tailwindcss-twin/issues/29), so lightyen can look into it 👍

**Reply by locona** [reply]
oh, thanks!!

**Comment by xlboy** [comment]
About the feature "`Apply variants to multiple classes at once with variant groups`"

How should it be configured to get a hint in `VSCode - Tailwind CSS IntelliSense` ?

**Comment by b0o** [comment]
For Neovim support, I translated the recommended config to Lua for use with lspconfig:

```lua
require('lspconfig').tailwindcss.setup {
  settings = {
    scss = { validate = false },
    editor = {
      quickSuggestions = { strings = true },
      autoClosingQuotes = 'always',
    },
    tailwindCSS = {
      experimental = {
        classRegex = {
          'tw`([^`]*)', -- tw`...`
          'tw="([^"]*)', -- <div tw="..." />
          'tw={"([^"}]*)', -- <div tw={"..."} />
          'tw\\.\\w+`([^`]*)', -- tw.xxx`...`
          'tw\\(.*?\\)`([^`]*)', -- tw(Component)`...`
        },
      },
      includeLanguages = {
        typescript = 'javascript',
        typescriptreact = 'javascript',
      },
    },
  },
}
```

**Comment by massoudsalem** [comment]
When using this with groups, IntelliSense gives a wrong warning:
![image](https://github.com/ben-rogerson/twin.macro/assets/36207910/7ff1268d-9c16-4bff-b811-73d6d0daceea)

**Reply by john-twigg-ck** [reply]
The project 

https://github.com/lightyen/vscode-tailwindcss-twin
 us now archived and out of date.

the errors its producing are becuase the underlying types project is itself out of date:
https://github.com/lightyen/tailwind-types

BOO!

**Comment by DominickVale** [comment]
Has anybody managed to get all auto completions working on neovim? Especially the css`` lang injection.
